### PR TITLE
TablePartitionWorker

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/warehouse/workers/db/AclSnapshotDaoImpl.java [hunk 1]
```diff
@@ -104,7 +104,11 @@ public void setValues(PreparedStatement ps, int i)
 					throws SQLException {
 				AclSnapshot snapshot = batch.get(i);
 				ps.setLong(1, snapshot.getTimestamp());
-				ps.setLong(2, Long.parseLong(snapshot.getId()));
+				String id = snapshot.getId().toLowerCase();
+				if (id.startsWith("syn")) {
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/warehouse/workers/db/AclSnapshotDaoImpl.java]: move to util.

File: src/main/java/org/sagebionetworks/warehouse/workers/db/TableCreatorImpl.java [hunk 2]
```diff
@@ -13,6 +20,15 @@
...
 	private DateTime endDate;
+	public static final String CHECK_PARTITION = "SELECT COUNT(*) "
+			+ "FROM INFORMATION_SCHEMA.PARTITIONS "
+			+ "WHERE TABLE_NAME = ? AND PARTITION_NAME = ?";
+	public static final String ALL_PARTITIONS = "SELECT PARTITION_NAME "
+			+ "FROM INFORMATION_SCHEMA.PARTITIONS "
+			+ "WHERE TABLE_NAME = ?";
+	public static final String ADD_PARTITION = "ALTER TABLE %1$S "
+			+ "ADD PARTITION (PARTITION %2$S VALUES LESS THAN (%3$d))";
+	public static final String DROP_PARTITION = "ALTER TABLE `%1$S DROP PARTITION %2$S";
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/warehouse/workers/db/TableCreatorImpl.java:32]: `?

File: src/main/java/org/sagebionetworks/warehouse/workers/utils/PartitionUtil.java [hunk 3]
```diff
@@ -11,30 +17,55 @@
 		MONTH
 	}
 
-	private final static String PARTITION_BY_RANGE = "PARTITION BY RANGE";
-	private final static String PARTITION_LESS_THAN = "PARTITION %1$s VALUES LESS THAN (%2$s)";
-	private final static String PARTITION_PREFIX = "_P";
+	public final static String PARTITION_BY_RANGE = "PARTITION BY RANGE";
+	public final static String PARTITION_LESS_THAN = "PARTITION %1$s VALUES LESS THAN (%2$s)";
+	public final static String PARTITION_NAME_PATTERN = "%1$S_%2$d_%3$d_%4$d";
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/warehouse/workers/utils/PartitionUtil.java:22]: pad zero no comparator

File: src/test/java/org/sagebionetworks/warehouse/workers/db/TableCreatorImplTest.java [hunk 4]
```diff
@@ -67,4 +66,19 @@ public void createTableBasedOnTableConfigPartitionFalseTest() {
 		assertTrue(argument.getValue().contains(TABLE_ACL_SNAPSHOT));
 		assertFalse(argument.getValue().contains(PartitionUtil.PARTITION));
 	}
+
+	@Test
+	public void testPartitionDoesNotExist() {
+		String partitionName = String.format(PartitionUtil.PARTITION_NAME_PATTERN, TABLE_ACCESS_RECORD, 2015, 1, 1);
+		creator.doesPartitionExist(TABLE_ACCESS_RECORD, partitionName);
+		Mockito.verify(mockTemplate).queryForLong(TableCreatorImpl.CHECK_PARTITION, TABLE_ACCESS_RECORD, partitionName);
+	}
```
Comment (john-hill) [inline src/test/java/org/sagebionetworks/warehouse/workers/db/TableCreatorImplTest.java]: duplicate test

File: src/main/java/org/sagebionetworks/warehouse/workers/db/WarehouseWorkersStateDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+			+ ") VALUES (?)";
+
+	private static final String GET_STATE = "SELECT "
+			+ COL_WAREHOUSE_WORKERS_STATE_STATE
+			+ " FROM "
+			+ TABLE_WAREHOUSE_WORKERS_STATE
+			+ " WHERE "
+			+ COL_WAREHOUSE_WORKERS_STATE_TIMESTAMP
+			+ " = (SELECT MAX("
+			+ COL_WAREHOUSE_WORKERS_STATE_TIMESTAMP
```
Comment (kimyen) [inline src/main/java/org/sagebionetworks/warehouse/workers/db/WarehouseWorkersStateDaoImpl.java:35]: could return more than 1 value

**Pull request merged**